### PR TITLE
[libc++] Allow testing Apple's system library as it is installed

### PR DESCRIPTION
**Pull request by ldionne**
In order to test libc++ under the "Apple System Library" configuration, we need to run the tests using DYLD_LIBRARY_PATH. This is required because libc++ gets an install_name of /usr/lib when built as a system library, which means that we must override the copy of libc++ used by the whole process. This effectively reverts 2cf2f1b, which was the wrong solution for the problem I was having.

Of course, this assumes that the just-built libc++ is sufficient to replace the system library, which is not actually the case out-of-the-box. Indeed, the system library contains a few symbols that are not provided by the upstream library, leading to undefined symbols when replacing the system library by the just-built one.

To solve this problem, we separately build shims that provide those missing symbols and we manually link against them when we build executables in the tests. While this is somewhat brittle, it provides a localized and unintrusive way to allow testing the Apple system configuration in an upstream environment, which has been a frequent request.

### CODE REVIEW

**Comment by ldionne** [comment]
I ran this locally and it seemed happy. If this works out, I think this is a better (more localized) alternative to #98070.

**Review by mordante — APPROVED** [review]
I really like how this patch turns out. LGTM but I have one comment.

File: libcxx/utils/ci/apple-install-libcxx.sh [hunk 1]
```diff
@@ -114,8 +114,13 @@ for arch in ${architectures}; do
     #       Then LLVM would guess the LLVM_DEFAULT_TARGET_TRIPLE properly and we wouldn't have to specify it.
     target=$(xcrun clang -arch ${arch} -xc - -### 2>&1 | grep --only-matching -E '"-triple" ".+?"' | grep --only-matching -E '"[^ ]+-apple-[^ ]+?"' | tr -d '"')
 
-    step "Building libc++.dylib and libc++abi.dylib for architecture ${arch}"
     mkdir -p "${build_dir}/${arch}"
+
+    step "Building shims to make libc++ compatible with the system libc++ on Apple platforms when running the tests"
+    shims_library="${build_dir}/${arch}/apple-system-shims.a"
+    xcrun clang++ -c -std=c++2b -target ${target} "${llvm_root}/libcxxabi/src/vendor/apple/shims.cpp" -static -o "${shims_library}"
```
[the remark this answers is not in the document]
Comment (ldionne) [reply libcxx/utils/ci/apple-install-libcxx.sh:121]: The values don't need to match exactly. I will add a comment mentioning that we don't have to match them.

**Comment by trcrsired** [comment]
Hello. Can you tell me what toggle i can use for building libcxxabi/src/vendor/apple/shims.cpp ? I am doing cross compiling from linux to mac

**Pull request merged**